File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int s_len = s.length();
        if (s_len <= 1) {
            return s_len;
        }
        
        int head = 0, index = 0, ret = 0;
        Map<Character, Integer> map = new HashMap<Character, Integer>();

        for (int i = 0; i < s_len;) {
            char c = s.charAt(i);
            if (map.get(c) != null) {
                ret = Math.max(i - head, ret);
                head = Math.max(head, map.get(c));

                if (s_len - head < ret) {
                    return ret;
                }
            }
            map.put(c, ++i);
        }
        return Math.max(s_len - head, ret);
    }
}

File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
 var lengthOfLongestSubstring = function(s) {
    var s_len = s.length;
    if (s_len <= 1)
        return s_len;

    var head = 0, ret = 0, map = {};
    
    for (var index = 0; index < s_len;) {
        var c = s.charAt(index);
        if (map[c] != undefined) {
            ret = Math.max(index - head, ret);
            head = Math.max(head, map[c]);
        }
        map[c] = ++index;
    }
    return Math.max(s_len - head, ret);    
};